src/test/java/underlay/MockUnderlayTest.java

```diff
@@ -0,0 +1,70 @@
+package underlay;
+
+import blockchain.LightChainNode;
+import blockchain.Parameters;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import underlay.mock.MockUnderlay;
+import underlay.requests.lightchain.GetModeRequest;
+import underlay.requests.lightchain.GetPublicKeyRequest;
+import underlay.requests.skipgraph.PingRequest;
+import underlay.responses.EmptyResponse;
+import util.Const;
+import util.Util;
+
+import java.io.FileNotFoundException;
+import java.rmi.RemoteException;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static underlay.responses.BooleanResponse.BooleanResponseOf;
+import static underlay.responses.PublicKeyResponse.PublicKeyResponseOf;
+
+/**
+ * To test the application layer of LightChain, we need to mock out the RMI underlay implementation.
+ */
+public class MockUnderlayTest {
+    private static MockUnderlay mockUnderlay1;
+    private static MockUnderlay mockUnderlay2;
+    static LightChainNode node1;
+    static LightChainNode node2;
+    private static Parameters params;
+    private static int port1 = 7050;
+    private static int port2 = 7051;
+    private static String IP = Util.grabIP();
+    private static String address1 = IP + ":" + port1;
+    private static String address2 = IP + ":" + port2;
+
+
+    @BeforeAll
+    static void init() {
+        params = new Parameters();
+        mockUnderlay1 = new MockUnderlay(address1);
+        node1 = new LightChainNode(params, port1, Const.DUMMY_INTRODUCER, true, mockUnderlay1);
+        mockUnderlay2 = new MockUnderlay(address2);
+        node2 = new LightChainNode(params, port2, node1.getAddress(), false, mockUnderlay2);
+    }
+
+
+    /**
+     * When both underlays A and B are up and ready, their exchanged messages are delivered to each other intact.
+     * We test different messages to make sure all of them are sent correctly
+     * We compare the results with the actual values returned by the nodes themselves
+     * @throws RemoteException
+     * @throws FileNotFoundException
+     */
+    @Test
+    void sendMessageTest() throws RemoteException, FileNotFoundException {
```

Comment:
Please add more `assertEqual`s involving the `SkipGraphNode` methods, e.g., `SearchByNumIDRequest`.

Reply:
Implemented in d0c717b428fdc07d6ea9ac45a768739880d91858